raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,164 @@
+        this.stateFile = stateFile;
+        cachedState = loadState(stateFile, metadataLog);
+        log.info("Loaded quorum state from local: {}", cachedState);
+    }
+
+    private QuorumStateData loadState(File file, final ReplicatedLog metadataLog) throws IOException {
+        try (final BufferedReader reader = Files.newBufferedReader(file.toPath())) {
+            final ObjectMapper objectMapper = new ObjectMapper();
+            final ObjectNode dataObject = (ObjectNode) objectMapper.readTree(readLine(reader));
+            final short dataVersion = dataObject.get(DATA_VERSION).shortValue();
+
+            QuorumStateData data = new QuorumStateData();
+            data.fromJson(dataObject, dataVersion);
+            return data;
+        } catch (final NoSuchFileException e) {
+            return recoverStateFromLog(file, metadataLog);
+        } catch (final EOFException e) {
+            delete();
+            log.info("Local quorum file corrupted, deleting {}", file.toPath());
+            return recoverStateFromLog(file, metadataLog);
+        }
+    }
+
+    private QuorumStateData recoverStateFromLog(File file, final ReplicatedLog metadataLog) throws IOException {
+        QuorumStateData stateFromMetadataLog = new QuorumStateData();
+        incorporateElectionState(stateFromMetadataLog,
+            ElectionState.withUnknownLeader(metadataLog.lastFetchedEpoch()));
+        writeElectionStateToFile(file, stateFromMetadataLog);
+        return stateFromMetadataLog;
+    }
+
+    /**
+     * Reads the election state from cached quorum data.
+     */
+    @Override
+    public ElectionState readElectionState() {
+        return new ElectionState(
+            cachedState.leaderEpoch(),
+            cachedState.leaderId() == UNKNOWN_LEADER_ID ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.leaderId()),
+            cachedState.votedId() == NOT_VOTED ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.votedId()));
+    }
+
+    @Override
+    public void writeElectionState(ElectionState latest) throws IOException {
+        incorporateElectionState(cachedState, latest);
+        writeElectionStateToFile(stateFile, cachedState);
+    }
+
+    private void incorporateElectionState(QuorumStateData state, ElectionState latest) {
+        state.setLeaderEpoch(latest.epoch)
+            .setVotedId(latest.hasVoted() ? latest.votedId() : NOT_VOTED)
+            .setLeaderId(latest.hasLeader() ? latest.leaderId() : UNKNOWN_LEADER_ID);
+    }
+
+    private void writeElectionStateToFile(final File stateFile, QuorumStateData state) throws IOException  {
+        final File temp = new File(stateFile.getAbsolutePath() + ".tmp");
+        log.trace("Writing tmp quorum state {}", temp.getAbsolutePath());
+        final FileOutputStream fileOutputStream = new FileOutputStream(temp);
+
+        try (final BufferedWriter writer = new BufferedWriter(
```

Comment:
Does `BufferedWriter.close()` also close the underlying writer? Looking at the code, it seems not.

Reply:
Will call writer.close explicitly

Reply:
In fact, is that necessary? I think once we flush the data to file and do the atomic swap, the underlying writer should no longer be useful?

Reply:
Just for cleanness, it is still preferable to close the writer object.

Reply:
My point was that closing the writer would not close the file output stream we created above. If we don't close the descriptor, it is a leak. The usual way to write these kinds of chains is to put them all in the `try`. For example:
```
    try (
        java.util.zip.ZipFile zf =
             new java.util.zip.ZipFile(zipFileName);
        java.io.BufferedWriter writer = 
            java.nio.file.Files.newBufferedWriter(outputFilePath, charset)
    ) {
```